### PR TITLE
Refactor rule_instances_controller and implement move_up/down methods

### LINKED ISSUE #21: Rule instance up/down arrows don't always work
Add a few instances, delete a few, then try to move some around.

### PR DESCRIPTION
**Pull request by daledavies**

### CODE REVIEW

File: classes/local/rule_instances_controller.php [hunk 1]
```diff
@@ -44,90 +44,179 @@
...
+        global $DB;
+        // We'll possibly be committing a number of new records and updates
+        // to the databse so best to create a new transaction.
+        $transaction = $DB->start_delegated_transaction();
+        // Iterate over the internal rule insatnces list, determine if we have
+        // anything to commit and perform the appropriate database operation.
+        foreach ($this->ruleinstancesinternal as $internal) {
+            // New instances.
+            if (!isset($internal->id)) {
+                $internal->id = $DB->insert_record('tool_registrationrules', $internal);
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php:84]: There's a parameter `bulk: true` to indicate bulk inserts. I don't think we've ever get to the use-case of inserting several hundred rules at once, but we could set it.

File: classes/local/rule_instances_controller.php [hunk 2]
```diff
@@ -44,90 +44,179 @@
...
+                $this->ruleinstances[$internal->id] = $internal;
+            }
+            // Modified instances.
+            if (isset($internal->modified)) {
+                unset($internal->modified);
+                $DB->update_record('tool_registrationrules', $internal);
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php:90]: Like for `$DB->insert_record()` we can also set a bulk parameter here.

File: classes/local/rule_instances_controller.php [hunk 3]
```diff
@@ -44,90 +44,179 @@
...
+                $this->ruleinstances[$internal->id] = $internal;
+            }
+            // Deleted instances.
+            if (isset($internal->deleted)) {
+                $DB->delete_records('tool_registrationrules', ['id' => $internal->id]);
+                unset($this->ruleinstances[$internal->id]);
+            }
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php:97]: It'd break your design here, but `delete_records()` is more efficient, if it deletes multiple records at once. We could collect their IDs and use `delete_records_list()` as example.

**Comment by phager-at** [comment]
I'm wondering if we should've implemented [CoS pattern](https://refactoring.guru/design-patterns/chain-of-responsibility) for our rule instances (for the rule checker as well as here for the controller) from the start. Or even gone as far as using a double linked list for them. :thinking:

I've thought about your implementation and the only other reasonable option I've come up with is to use `array_keys($this->ruleinstances)` to get an ordered list of our rule instance ids and we could also use `array_flip()` to reverse it to a map from "sort order" (or indexes 0…n) to our instance ids. There we could more simply get our instance ids' sort order numbers and simply get the instance id at index n-1 or n+1. The only downside is, we'd have to keep the ordered array of instanceids in sync with our `ruleinstancesinternal` array. It wouldn't mind me, if our sort order starts at 0 like an array index.

```php
$mapordertoinstanceid = array_keys($this->ruleinstances);
$mapinstanceidtoorder = array_flip($mapordertoinstanceid);

// Given instances X, Y and Z (in that order):
$orderofY = $mapordertoinstanceid[Y];
// if $oderofY > 0 we have a previous element, else we don't!
$prevInstanceId = $orderofY > 0 ? $mapinstanceidtoorder[$orderofY-1] : null;
// If $orderofY is already the last key in $mapordertoinstanceid, we don't have a following element!
$nextInstanceId = $orderofY !== array_key_last($mapordertoinstanceId) ? $mapinstanceidtoorder[$orderofY+1] : null;
```

**Comment by phager-at** [comment]
(should've put this in the review submission box)

As far as I see it, the implementation using the `commit()` method would be perfect for handling multiple actions in one request. For now it adds a bit of complexity to the code. But it's nothing we'd have to absolutely change.

**Comment by daledavies** [comment]
> (should've put this in the review submission box)
> 
> As far as I see it, the implementation using the `commit()` method would be perfect for handling multiple actions in one request. For now it adds a bit of complexity to the code. But it's nothing we'd have to absolutely change.

I've taken this further in my next round of changes (not pushed yet) so that we can perform multiple actions without instantly committing them, and then call `commit()` manually later.

For example, note the addition of the `$commit` param here...

```
    public function update_instance(stdClass $formdata, bool $commit = true): void {
         $formdata->type = $this->ruleinstancesinternal[$formdata->id]->type;
        // Update default fields in record.
        foreach ($this->extract_instancedata($formdata) as $property => $value) {
            if ($property != 'type') {
                $this->ruleinstancesinternal[$formdata->id]->{$property} = $value;
            }
        }
        // Encode rule specific config data from the form and add to the instance record.
        $this->ruleinstancesinternal[$formdata->id]->other = $this->encode_instance_config($formdata);
        // Signify we have made a modification.
        $this->ruleinstancesinternal[$formdata->id]->modified = true;
        // Commit the update to the database now if required.
        if ($commit) {
            $this->commit();
        }
    }
```

File: classes/local/rule_instances_controller.php (hunk 1 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php:84]: Craziest thing is that it doesn't seem to be used for postgres or mysql!

Comment (daledavies) [reply classes/local/rule_instances_controller.php:84]: In fact, I can't see it used for any of the DB plugins. It's passed to things like `insert_record_raw()` but they don't use it.

File: classes/local/rule_instances_controller.php (hunk 3 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php:97]: I can see how this would work.

File: classes/local/rule_instances_controller.php (hunk 1 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php:84]: Looks like there's an `insert_records()` which takes an array of objects to insert into the table, I can see a way of using that in the same way I'm thinking to make changes to use an array of records to delete in one go with `delete_records()`.

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php:90]: Unfortunately I fear this would do nothing :(

**Comment by daledavies** [comment]
I have some ideas based on your feedback, I'll see what I can do :)

**Comment by daledavies** [comment]
Maybe we can just drop the transaction and just do updates in a loop, then collect records or id's for insert_records() and delete_records() or delete_records_list()?

**Comment by daledavies** [comment]
Or we could keep the transaction and do all of the above within it, I added it just as a way to keep related changes atomic

**Comment by phager-at** [comment]
I think the transaction is good! I'd keep it there.

And on second thought, the other changes regarding collecting inserts and deletes. I think we can keep it as it is, we wont have that many records here to process…

**Comment by daledavies** [comment]
> I'm wondering if we should've implemented [CoS pattern](https://refactoring.guru/design-patterns/chain-of-responsibility) for our rule instances (for the rule checker as well as here for the controller) from the start. Or even gone as far as using a double linked list for them. 🤔
> 
> I've thought about your implementation and the only other reasonable option I've come up with is to use `array_keys($this->ruleinstances)` to get an ordered list of our rule instance ids and we could also use `array_flip()` to reverse it to a map from "sort order" (or indexes 0…n) to our instance ids. There we could more simply get our instance ids' sort order numbers and simply get the instance id at index n-1 or n+1. The only downside is, we'd have to keep the ordered array of instanceids in sync with our `ruleinstancesinternal` array. It wouldn't mind me, if our sort order starts at 0 like an array index.
> 
> ```
> $mapordertoinstanceid = array_keys($this->ruleinstances);
> $mapinstanceidtoorder = array_flip($mapordertoinstanceid);
> 
> // Given instances X, Y and Z (in that order):
> $orderofY = $mapordertoinstanceid[Y];
> // if $oderofY > 0 we have a previous element, else we don't!
> $prevInstanceId = $orderofY > 0 ? $mapinstanceidtoorder[$orderofY-1] : null;
> // If $orderofY is already the last key in $mapordertoinstanceid, we don't have a following element!
> $nextInstanceId = $orderofY !== array_key_last($mapordertoinstanceId) ? $mapinstanceidtoorder[$orderofY+1] : null;
> ```

Can we leave this for a future improvement? Unless you think it would offer significant gains.

**Comment by phager-at** [comment]
> > I'm wondering if we should've implemented [CoS pattern](https://refactoring.guru/design-patterns/chain-of-responsibility) for our rule instances (for the rule checker as well as here for the controller) from the start. Or even gone as far as using a double linked list for them. 🤔
> > I've thought about your implementation and the only other reasonable option I've come up with is to use `array_keys($this->ruleinstances)` to get an ordered list of our rule instance ids and we could also use `array_flip()` to reverse it to a map from "sort order" (or indexes 0…n) to our instance ids. There we could more simply get our instance ids' sort order numbers and simply get the instance id at index n-1 or n+1. The only downside is, we'd have to keep the ordered array of instanceids in sync with our `ruleinstancesinternal` array. It wouldn't mind me, if our sort order starts at 0 like an array index.
> > ```
> > $mapordertoinstanceid = array_keys($this->ruleinstances);
> > $mapinstanceidtoorder = array_flip($mapordertoinstanceid);
> > 
> > // Given instances X, Y and Z (in that order):
> > $orderofY = $mapordertoinstanceid[Y];
> > // if $oderofY > 0 we have a previous element, else we don't!
> > $prevInstanceId = $orderofY > 0 ? $mapinstanceidtoorder[$orderofY-1] : null;
> > // If $orderofY is already the last key in $mapordertoinstanceid, we don't have a following element!
> > $nextInstanceId = $orderofY !== array_key_last($mapordertoinstanceId) ? $mapinstanceidtoorder[$orderofY+1] : null;
> > ```
> 
> Can we leave this for a future improvement? Unless you think it would offer significant gains.

Nope, it was just meant as a creative input, and the only other way to get to our desired result I could think of today. :-)

**Review by phager-at — APPROVED** [review]
LGTM

**Pull request merged**